src/parcsr_ls/par_relax.c

```diff
@@ -1409,3 +1466,150 @@ hypre_BoomerAMGRelaxKaczmarz( hypre_ParCSRMatrix *A,
+      }
+   }
+
+   for (i = 0; i < num_rows; i++) /* Run the smoother */
+   {
+      HYPRE_Complex res = 0.0;
+      for (jj = A_diag_i[i]; jj < A_diag_i[i+1]; jj++)
+      {
+         ii = A_diag_j[jj];
+         if (ii < i)
+         {
+            res -= (A_diag_data[jj] / A_diag_data[A_diag_i[ii]]) * Vtemp_data[ii];
+         }
+      }
+      u_data[i] += (Vtemp_data[i] + omega*res) / A_diag_data[A_diag_i[i]];
+   }
+
+   return hypre_error_flag;
+}
+
+HYPRE_Int
+hypre_BoomerAMGRelax11TwoStageGaussSeidel( hypre_ParCSRMatrix *A,
+                                           hypre_ParVector    *f,
+                                           HYPRE_Int          *cf_marker,
+                                           HYPRE_Int           relax_points,
+                                           HYPRE_Real          relax_weight,
+                                           HYPRE_Real          omega,
+                                           hypre_ParVector    *u,
+                                           hypre_ParVector    *Vtemp,
+                                           hypre_ParVector    *Ztemp )
+{
+#if defined(HYPRE_USING_CUDA)
+   //HYPRE_ExecutionPolicy exec = hypre_GetExecPolicy2( hypre_VectorMemoryLocation(x), hypre_VectorMemoryLocation(b) );
+   //RL: TODO back to hypre_GetExecPolicy2 later
+   HYPRE_ExecutionPolicy exec = HYPRE_EXEC_DEVICE;
+
+   if (exec == HYPRE_EXEC_DEVICE)
+   {
+      hypre_BoomerAMGRelaxTwoStageGaussSeidelDevice(A, f, relax_weight, omega, u, Vtemp, Ztemp, 1);
+   }
+   else
+#endif
+   {
+      hypre_BoomerAMGRelax11TwoStageGaussSeidelHost(A, f, relax_weight, omega, u, Vtemp);
+   }
+
+   return hypre_error_flag;
+}
+
+HYPRE_Int
+hypre_BoomerAMGRelax12TwoStageGaussSeidelHost( hypre_ParCSRMatrix *A,
+                                               hypre_ParVector    *f,
+                                               HYPRE_Real          relax_weight,
+                                               HYPRE_Real          omega,
+                                               hypre_ParVector    *u,
+                                               hypre_ParVector    *Vtemp )
+{
+   hypre_CSRMatrix *A_diag      = hypre_ParCSRMatrixDiag(A);
+   HYPRE_Int        num_rows    = hypre_CSRMatrixNumRows(A_diag);
+   HYPRE_Real      *A_diag_data = hypre_CSRMatrixData(A_diag);
+   HYPRE_Int       *A_diag_i    = hypre_CSRMatrixI(A_diag);
+   HYPRE_Int       *A_diag_j    = hypre_CSRMatrixJ(A_diag);
+   hypre_Vector    *Vtemp_local = hypre_ParVectorLocalVector(Vtemp);
+   HYPRE_Complex   *Vtemp_data  = hypre_VectorData(Vtemp_local);
+   hypre_Vector    *u_local     = hypre_ParVectorLocalVector(u);
+   HYPRE_Complex   *u_data      = hypre_VectorData(u_local);
+   HYPRE_Int        i, jj, ii;
+
+   /* Need to check that EVERY diagonal is nonzero first. If any are, throw exception */
+   for (i = 0; i < num_rows; i++)
+   {
+      if (A_diag_data[A_diag_i[i]] == 0.0)
+      {
+         hypre_error_in_arg(1);
```

Comment:
@PaulMullowney Please check if this error handling makes sense. Thanks!

Reply:
Also fine.